### DISCUSSION TITLE
Discussion on improving Cljdoc infrastructure & operations

### DISCUSSION THREAD

**Discussion by rvalentini** [cljdoc/cljdoc#505] [Ideas]
Here we can gather ideas on how to improve the current infrastructure setup and simplify operations.  

Discussion results should be documented in form of ADRs and actionable issues.  

Status Quo documentation: https://github.com/cljdoc/cljdoc/tree/master/ops

**Comment by rvalentini** [comment]
### Topic: Simplify the Cljdoc deployment setup

Documentation: https://github.com/cljdoc/cljdoc/blob/master/doc/adr/0017-use-nomad-for-deployment.md

**Reply by rvalentini** [reply]
@lread  suggested AWS ElasticBeanstak (EBS) as an alternative to the current setup (DigitalOcean Droplet + Nomad for orchestration + Traefik for load balancing).

This would result in fewer moving parts that we have to take care of and could streamline the deployment process.

Cost model for EBS: The orchestration service itself is for free, you pay for the underlying resource usage. In our case, this would be the EC2 instance running the Cljdoc server and the attached load balancer. Depending on the traffic, the load balancer can become quite pricey, though. With some traffic statistics, I could look up the monthly costs for this option.

**Reply by martinklepsch** [reply]
[Vercel](https://vercel.com/) could be another option here, they have a tier for $20 with integrated edge caching and other goodies. (They seem to sponsor some projects if you include [some advertising](https://vercel.com/support/articles/can-vercel-sponsor-my-open-source-project)).

**Reply by martinklepsch** [reply]
Google Cloud Run is another interesting option that we could consider. It has a free tier/budget that should cover most expenses depending on usage.

**Reply by rvalentini** [reply]
Had a quick look at Vercel and found this limitation: https://vercel.com/support/articles/does-vercel-support-docker-deployments That's a deal-breaker since the cljdoc service is currently bundled as Docker image, right?

**Reply by lread** [reply]
@rvalentini I think maybe we can keep options open at this stage.

I'm not sure that deploying a docker image is a hard requirement.
Deploying a jar might be ok too.

That said, we do support running cljdoc locally via a docker image. 
So we'd have to think about that use case too.

**Reply by lread** [reply]
Might be on the obvious side, but probably worth exploring what problems are trying to solve. 
Some symptoms:
- It is hard for newcomers to understand ops because there are quite a few moving parts and technologies at play.
- This makes it hard to enlist any help cljdoc ops help from the community.
- Because cljdoc ops is not a full-time job, these tools and technologies are not kept up to date, which makes them harder to maintain. Some of these tools have undergone significant revisions involving such things as file format changes.
- Scripting around these tools is also not kept up to date for the same reason.
- Scripting around these tools is on the complex side. It involves orchestrating all of the moving parts we are managing ourselves.

@martinklepsch might have more to add here.

**Reply by martinklepsch** [reply]
Ah, I didn't realize Vercel moved away completely from Docker! So yeah, serverless style deployments are probably not a good fit with cljdoc's long-ish startup time.

**Comment by rvalentini** [comment]
### Topic: Move the application state away from the machine running the service

Documentation: https://github.com/cljdoc/cljdoc/blob/master/doc/adr/0013-move-to-sqlite-for-storage.md

**Reply by martinklepsch** [reply]
https://planetscale.com/ seems like an interesting option for this. We'd probably need to make some small adjustments to the codebase to support MySQL but shouldn't be too hard. 

Similarly we could use hosted Postgres from DigitalOcean or Exoscale (who use Clojure so might be willing to sponsor, cc @pyr).

**Reply by pyr** [reply]
Happy to sponsor infrastructure for clj-doc, @martinklepsch you can reach me by email (same handle at exoscale.com)

**Reply by martinklepsch** [reply]
http://pgloader.io/ Seems like a good option to move state from sqlite into postgres.

**Reply by rvalentini** [reply]
Had a look at pgloader today. I tried to load a cljdoc backup into a local Postgres and ran into this issue: https://github.com/dimitri/pgloader/issues/1183 As a workaround I downgraded to an older Postgres version (`14.1` -> `13.4`). The migration started after that but halfway through the `builds` table the import stopped and the process was hanging. So first impression: not plug & play, unfortunately.

**Reply by lread** [reply]
@rvalentini good to know.
The cljdoc db is on the simple side.  
I expect we could whip up a little clojure program to to the loading without too much effort.
My guess is that this will be a one-time load?
And also that maybe we'd be ok with slightly stale data for the transition.
Be interesting to know how long the load would take, but we'll learn that, I suppose!

**Reply by martinklepsch** [reply]
> And also that maybe we'd be ok with slightly stale data for the transition.

I think that's ok and we could always check which clojars releases were made during the transition and trigger builds for those.

**Reply by lread** [reply]
Another thing to think about is how this fits in with database migrations.
We currently use ragtime.
Is it unrealistic to treat moving to a different db tech as a ragtime migration?

**Reply by lread** [reply]
Are there specific impacts we want to test/explore?
For example, I wonder what the impact of fetching/saving to a remote store will be.
- Any latency concerns?
- Any data transfer $ cost concerns? 

Might be totally fine, but also interesting to explore.

**Reply by rvalentini** [reply]
> Another thing to think about is how this fits in with database migrations. We currently use ragtime. Is it unrealistic to treat moving to a different db tech as a ragtime migration?

Interesting point! I see moving to a different DB tech rather independent from the actual DB schema. The best-case scenario would be that we can use the existing migration scripts without any changes on a new Postgres e.g., right? Maybe this could also be part of the little Clojure program that you mentioned to get the data transitioned.

**Reply by lread** [reply]
> The best-case scenario would be that we can use the existing migration scripts without any changes on a new Postgres e.g., right? Maybe this could also be part of the little Clojure program that you mentioned to get the data transitioned.

Haven't really thought it through yet. 
But the idea is that a migration would transfer our state held in SQLite to some other db tech (maybe Postgres?). 
Since it seems kind of silly to carry around SQLite after the successful migration, we'd probably squash our migrations to move forward with our new db tech only.
Unless there are other cljdoc installations I don't know about. Then we might carry around SQLite only to support their migration paths as well?

**Reply by rvalentini** [reply]
> > The best-case scenario would be that we can use the existing migration scripts without any changes on a new Postgres e.g., right? Maybe this could also be part of the little Clojure program that you mentioned to get the data transitioned.
> 
> Haven't really thought it through yet. But the idea is that a migration would transfer our state held in SQLite to some other db tech (maybe Postgres?). Since it seems kind of silly to carry around SQLite after the successful migration, we'd probably squash our migrations to move forward with our new db tech only. Unless there are other cljdoc installations I don't know about. Then we might carry around SQLite only to support their migration paths as well?

What I like about the idea is that any change to the DB is fully covered by a migration and we don't have to do any ad-hoc/manual operations. 

On the downside, it might also introduce some additional complexity since DB schema evolution and the data transition sound like somewhat different concerns to me. If in the end there is a set of migrations A (before the transition), and another set of migrations B (after the transition), this approach doesn't sound so promising to me anymore. If there is a way to keep the migration chain "intact" and just continue with the new DB tech it would be nice though.

**Reply by lread** [reply]
> On the downside, it might also introduce some additional complexity since DB schema evolution and the data transition sound like somewhat different concerns to me. If in the end there is a set of migrations A (before the transition), and another set of migrations B (after the transition), this approach doesn't sound so promising to me anymore. If there is a way to keep the migration chain "intact" and just continue with the new DB tech it would be nice though.

We should get confirmation from @martinklepsch if he knows of any other folks running their own cljdoc server instance. Maybe we only have to care about cljdoc.org.

The way I'm imagining this idea, the migration path would not allow users to stick with SQLite.
Let's say you are at migration 8 on SQLite.
Let's imagine migrations 9 and 10 are still on SQLite.
But it is migration 11 that converts you over to Postgres (which would fail if any pre-requisite tools were not in place).
That's it, you are now on Postgres.
We might keep SQLite JDBC drivers around to support folks who have not migrated yet.
We might decide to squash migrations to start at migration 11 (which would not support migrating SQLite->Postgres anymore).

**Reply by lread** [reply]
@rvalentini had a good chat exploring migrations and I've come to the conclusion that while my idea sounds good in the abstract, it just might not be practical. In short:
- It might be conceptually difficult for migration tools to deal with.
- It also might be horribly inefficient and/or unnecessarily costly.

So... do what is practical.  If re-starting at migration 0 and doing a bulk import is more practical, then that's totally fine.

**Reply by rvalentini** [reply]
> Are there specific impacts we want to test/explore? For example, I wonder what the impact of fetching/saving to a remote store will be.
> 
>     * Any latency concerns?
> 
>     * Any data transfer $ cost concerns?
> 
> 
> Might be totally fine, but also interesting to explore.

I gathered some information to answer these questions:

Concerning cost:
* Although we will receive Exoscale sponsoring :) I had a look at their pricing models. We would fit the smallest subscription plan called "Hobbyist", which comes at 38€ / month for Postgres DBaaS. There is no additional cost for networking. 
* On DigitalOcean side there might be additional costs for outbound traffic when we reach a certain threshold (1000GiB/month), but currently, our traffic is far below that threshold, so also no concerns here I would say. 

Concerning latency:
* Our DigialOcean droplet runs in one of the Amsterdam regions as far as I could see. 
* Exoscale offers multiple locations across Europe, of which Frankfurt would be the most favorable for us in this case. 
* When we don't touch the DigitalOcean droplet and choose Exoscale Frankfurt region for our DBaaS, I would estimate an additional network latency of 30ms for the round trip, plus some small amount of processing time. 
* When we move the droplet also to the Frankfurt region we could further cut down on the network latency if we need to.

So in theory I see no problems, we would have to try out to be sure though.

**Reply by lread** [reply]
Nice research @rvalentini! Thanks for digging in!

**Reply by rvalentini** [reply]
Had a great talk with @martinklepsch  about how to proceed further on this topic. To get a bit more concrete and explore possible implications, we want to try and connect Postgres as a new storage backend to the application.  

Another related question came up though: 
What about the SQLite Cache (https://github.com/cljdoc/cljdoc/blob/master/src/cljdoc/util/sqlite_cache.clj)? Does it make sense to also migrate that away from the local disk? As described in the ADR (https://github.com/cljdoc/cljdoc/blob/master/doc/adr/0018-caching-layer.md) the main purpose of it is to avoid network calls. Moving it to an external DBaaS together with the main SQLite DB seems to contradict that intent to some extent.   

One of the major benefits of the intended DBaaS migration would be to get rid of the local disk storage, which would open up more options to simplify the infrastructure setup. If the cache remains, and we use both Postgres and SQLite in parallel, the migration might not be so rewarding anymore?

**Reply by lread** [reply]
Awesome news @rvalentini, thanks for sharing!

Fwiw, I think the local cache is only currently used to cache poms.
Secondary benefits of a cache are sometimes:
- avoiding rate limits
- politeness (not overusing a 3rd party service) 
- continuing to operate for cached items when 3rd party service is down

Some popular typical cloud-based caching technologies are Redis and Memcached (or compatible) but I'm not sure if they fit our use case.

**Reply by martinklepsch** [reply]
I think it'd potentially be ok to wipe the cache on restarts/redeploys. We get the POM for every documentation request and as long as we only do it the first time documentation is loaded and then have it cached for a reasonable while (likely days as we don't deploy all that often) that seems ok to me?

**Comment by rvalentini** [comment]
### Documentation: Current Setup

To understand the current setup better, I tried to sketch the status quo of the infrastructure and what's going on from an operations perspective. Hope this will be helpful for anybody joining the discussion. This is my initial understanding from going over the config files & scripts, so if anything is wrong or missing, please tell and I will update the sketch.

**Reply by rvalentini** [reply]
![cljdoc_ops_v2 drawio](https://user-images.githubusercontent.com/20353357/152561277-d46a3ab4-1472-4f87-a423-90f4157c13be.png)

**Reply by lread** [reply]
Awesome! Would be interesting to explicitly draw in where the sqlite databases live.

**Reply by rvalentini** [reply]
Good idea, that would be the filesystem of the VM running Nomad and the two containers I think. I will add it to the sketch :+1:

**Comment by lread** [comment]
### Use case: I am cljdoc developer and need to run locally

Here we explore what this might mean in our new world.

**Comment by lread** [comment]
### Use case: I am library author and want to run locally to preview what my docs will look like

I'm not sure how many folks do this (I do but... yeah).

Current available paths are:
1. from docker
2. from source

**Comment by lread** [comment]
### Use case: I am a cljdoc developer and I'd like to have confidence that my changes will work in a production environment

Today, we get a reasonable level of confidence by running locally under docker.

**Comment by lread** [comment]
### Use case: I am a cljdoc developer and I want to work with a copy of the production db locally.

Currently on the easy side, an sqlite db is simply a file.

**Comment by rvalentini** [comment]
### Topic: Playground Cljdoc environment for experiments 

When we want to try out different infrastructure setups or migration possibilities, a second Cljdoc environment might be useful. Maybe we could also use the cljdoc.xyz domain here?

**Comment by lread** [comment]
### Topic: Logging Service
Stick with Sentry? Move to something else? Maybe papertrail?
Inspired by #513
